Mathlib/Tactic/Linter/Papercut.lean

```diff
@@ -0,0 +1,102 @@
+/-
+Copyright (c) 2024 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import Lean.Elab.Command
+import Lean.Linter.Util
+import Batteries.Data.List.Basic
+
+/-!
+# The papercut linter
+
+The papercut linter tries to ease new users into Lean.
+It flags certain design decisions that may be surprising or confusing.
+
+For instance,
+* `3 - 5 = 0` (`Nat`-subtraction is truncated subtraction),
+* `3 / 5 = 0` (`Nat`-division is rounded division),
+* `(3 : ℤ) / 5 = 0` (`Int`-division is rounded division),
+* `x / 0 = 0` for most types (division by `0` equals `0`).
+-/
+
+open Lean Elab Command
+
+namespace Mathlib.Linter.Papercut
+
+/-- `getPapercuts e` takes as input an expression `e` and returns an `Option MessageData`,
+representing a helpful message in case the expression `e` involves certain
+implementation details that may be confusing.
+
+For instance, it flags `Nat.sub`, `Nat.div`, `Int.div` and division by `0`.
+-/
+def getPapercuts (e : Expr) : Meta.MetaM (Option MessageData) := do
+  let expNat := mkConst ``Nat
+  let expInt := mkConst ``Int
+  let expZero := mkConst ``Nat.zero
+  match e.getAppFnArgs with
+    | (``HSub.hSub, #[n1, n2, n3, _, a, b]) =>
+      if ← [n1, n2, n3].allM (Meta.isDefEq · expNat) then
+        return some m!"Subtraction in ℕ is actually truncated subtraction: e.g. `1 - 2 = 0`!\n\
+                       This yields the 'expected' result only when you also prove the inequality\n\
+                       '{← Meta.ppExpr b} ≤ {← Meta.ppExpr a}'"
+      else return none
+    | (``HDiv.hDiv, #[n1, n2, n3, a, b, zero]) =>
+      let zer ← Meta.mkAppOptM ``OfNat.ofNat #[← instantiateMVars n3, expZero, none]
+      if ← Meta.isDefEq zer zero then
+        return some m!"Division by `0` is usually defined to be zero: e.g. `3 / 0 = 0`!\n\
+                       This is usually defined to be `0`, to avoid having to constantly prove \
+                       that denominators are non-zero."
+      else if ← [n1, n2, n3].allM (Meta.isDefEq · expNat) then
+        return some m!"Division in ℕ is actually the floor of the division: e.g. `1 / 2 = 0`!\n\
+                       This yields the 'expected' result only when you also prove that \
+                       '{← Meta.ppExpr b}' divides '{← Meta.ppExpr a}'"
+      else if ← [n1, n2, n3].allM (Meta.isDefEq · expInt) then
+        return some m!"\
+                    Division in ℤ is actually the floor of the division: e.g. `(1 : ℤ) / 2 = 0`!\n\
+                    This yields the 'expected' result only when you also prove that \
+                    '{← Meta.ppExpr b}' divides '{← Meta.ppExpr a}'"
+      else return none
+    | _ => return none
+
+/-- The "papercut" linter emits a warning on certain quirks of the basic
+definitions. -/
+register_option linter.papercut : Bool := {
+  defValue := false
```

Comment:
Let me also point out that, by default, this linter is inactive!

Reply:
I was going to ask about that actually. Given that this is off by default, but also seems meant to be helpful to *new* users who may not know what a linter is in the first place or that this one exists, what do you imagine is the workflow where someone ends up seeing these? They ask on Zulip and someone tells them to enable this linter?

Perhaps this is a broader topic better for elsewhere but perhaps there's room for an "easy mode / training wheel" switch which turns on this but potentially other guardrails, and new users are directed to start using Mathlib in that mode until they know they don't need it?

Reply:
Yes, what I had in mind is that someone is confused once, asks on Zulip and this linter is mentioned.

It would be great if options could be set more freely, so that you could have the option on by default when you clone mathlib, but you can change the default in CI/locally if you want to.

I think that making the linter on by default would cause more noise for people who already know what they are doing, than be beneficial to new users... if you can compare these two quantities!

Reply:
Actually, maybe I should run an experiment and see how many uses of `Nat`-subtraction and `Nat`-division there are in mathlib, to get a quantitative sense of what the extent of this could be.